crates/numcodecs-python/src/schema.rs

```diff
@@ -0,0 +1,486 @@
+            signature.push_str(&format!("{default}"));
+        } else if !parameter.required {
+            signature.push_str("=None");
+        }
+    }
+
+    if parameters.additional {
+        signature.push_str(", **kwargs");
+    }
+
+    signature
+}
+
+#[allow(clippy::too_many_lines)] // FIXME
+fn parameters_from_schema(schema: &Schema) -> Parameters {
+    if schema.as_bool() == Some(true) {
+        return Parameters {
+            named: Vec::new(),
+            additional: true,
+        };
+    }
+
+    let Some(schema) = schema.as_object() else {
+        return Parameters {
+            named: Vec::new(),
+            additional: false,
+        };
+    };
+
+    let mut parameters = Vec::new();
+
+    let required = match schema.get("required") {
+        Some(Value::Array(required)) => &**required,
+        _ => &[],
+    };
+
+    if let Some(Value::Object(properties)) = schema.get("properties") {
+        for (name, parameter) in properties {
+            parameters.push(Parameter {
+                name,
+                required: required
+                    .iter()
+                    .any(|r| matches!(r, Value::String(n) if n == name)),
+                default: parameter.get("default"),
+                docs: match parameter.get("description") {
+                    Some(Value::String(docs)) => Some(Cow::Borrowed(docs)),
+                    _ => None,
+                },
+            });
+        }
+    }
+
+    let mut additional = !matches!(schema.get("additionalProperties"), Some(Value::Bool(false)));
+
+    if let Some(Value::Array(variants)) = schema.get("oneOf") {
+        additional |= !matches!(schema.get("additionalProperties"), Some(Value::Bool(false)));
+
+        let mut variant_parameters = HashMap::new();
+
+        for (i, schema) in variants.iter().enumerate() {
+            let required = match schema.get("required") {
+                Some(Value::Array(required)) => &**required,
+                _ => &[],
+            };
+            let variant_docs = match schema.get("description") {
+                Some(Value::String(docs)) => Some(docs.as_str()),
+                _ => None,
+            };
+
+            if let Some(Value::Object(properties)) = schema.get("properties") {
+                for (name, parameter) in properties {
+                    let required = required
+                        .iter()
+                        .any(|r| matches!(r, Value::String(n) if n == name));
+                    let default = parameter.get("default");
+                    let r#const = parameter.get("const");
+                    let docs = match parameter.get("description") {
+                        Some(Value::String(docs)) => Some(Cow::Borrowed(docs.as_str())),
+                        _ => None,
+                    };
+
+                    match variant_parameters.entry(name) {
+                        Entry::Vacant(entry) => {
+                            let (docs, tag_docs) = match r#const {
```

Comment:
if the parameter looks like a tag, collect cross-variant docs, otherwise just use the parameter docs